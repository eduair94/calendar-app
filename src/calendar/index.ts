export * from './pages'
export * from './components/NavBar'
export * from './components/CalendarEvent'
export * from './components/CalendarModal'
export * from './components/FabAddNew'
export * from './components/FabDelete'