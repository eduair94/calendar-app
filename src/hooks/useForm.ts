import { useEffect, useMemo, useState } from 'react';

interface formCheckedValuesI {
    [key: string]: string | null
}

interface formStateI {
    [key: string]: string;
}

interface formValidationsI {
    [key: string]: [fn: (key:string) => boolean, errorMessage: string]
}

type IForm = {
    formState: formStateI; // Replace 'any' with the actual type
    onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void; // Replace 'any' with the actual type
    onResetForm: () => void;
    formValidation: formCheckedValuesI; // Replace 'any' with the actual type
    isFormValid: boolean;
} & formStateI & formCheckedValuesI;

export const useForm = ( initialForm = {}, formValidations: formValidationsI = {}): IForm => {
  
    const [ formState, setFormState ] = useState<formStateI>( initialForm );
    const [ formValidation, setFormValidation ] = useState<formCheckedValuesI>({});

    useEffect(() => {
        createValidators();
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [ formState ])

    useEffect(() => {
        setFormState( initialForm );
    }, [ initialForm ])
    
    
    const isFormValid = useMemo( () => {

        for (const formValue of Object.keys( formValidation )) {
            if ( formValidation[formValue] !== null ) return false;
        }

        return true;
    }, [ formValidation ])


    const onInputChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value } = target;
        setFormState({
            ...formState,
            [ name ]: value
        });
    }

    const onResetForm = () => {
        setFormState( initialForm );
    }

    const createValidators = () => {
        
        const formCheckedValues: formCheckedValuesI = {};
        
        for (const formField of Object.keys( formValidations )) {
            const [ fn, errorMessage ] = formValidations[formField];

            formCheckedValues[`${ formField }Valid`] = fn( formState[formField] ) ? null : errorMessage;
        }

        setFormValidation( formCheckedValues );
    }

    return {
        ...formState,
        formState,
        onInputChange,
        onResetForm,
        ...formValidation,
        isFormValid
    } as IForm
} 