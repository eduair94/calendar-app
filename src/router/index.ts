export * from './AppRouter'
export * from './AppRouter.interface'