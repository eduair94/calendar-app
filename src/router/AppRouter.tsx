import { Navigate, Route, Routes } from "react-router-dom";
import { AuthStatus } from ".";
import { LoginPage } from "../auth";
import { CalendarPage } from "../calendar";

export const AppRouter = () => {
  const authStatus: AuthStatus = AuthStatus.authenticated;
  return (
    <Routes>
      {authStatus === AuthStatus.notAuthenticated ? (
        <Route path="/auth/*" element={<LoginPage />} />
      ) : (
        <Route path="/*" element={<CalendarPage />} />
      )}

      <Route path="/*" element={<Navigate to="/auth/login" />} />
    </Routes>
  );
};
