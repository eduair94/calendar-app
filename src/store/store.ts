import { configureStore } from "@reduxjs/toolkit";
import { uiSlice } from "./ui";
import { calendarSlice } from "./calendar";

export const store = configureStore({
    reducer: {
        ui: uiSlice.reducer,
        calendar: calendarSlice.reducer
    },
    middleware(getDefaultMiddleware) {
        return getDefaultMiddleware({
            serializableCheck: false
        })
    },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch