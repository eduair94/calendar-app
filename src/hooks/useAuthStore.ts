import { useDispatch, useSelector } from "react-redux"
import { AuthUserI, RootState, clearErrorMesssage, onChecking, onLogin, onLogout, onLogoutCalendar } from "../store";
import { calendarApi } from "../api";
import { AxiosError } from "axios";

export const useAuthStore = () => {
    const { status, user: userE, errorMessage } = useSelector((state: RootState) => state.auth);
    const dispatch = useDispatch();
    const user = userE as AuthUserI;
    
    const startLogin = async({ email, password }: { email: string, password: string }) => {
        dispatch(onChecking());
        try {
            const {data} = await calendarApi.post('/auth', { email, password });
            localStorage.setItem('token', data.token);
            localStorage.setItem('token-init-date', new Date().getTime().toString());
            dispatch(onLogin({name: data.name, _id: data.uid}))
        } catch(error) {
            console.log("error", error);
            dispatch(onLogout("Invalid Email / Password"))
            setTimeout(()=> {
                dispatch(clearErrorMesssage())
            }, 10)
        }
    };

    const startRegister = async ({name, email, password} : {name: string; email:string; password:string}) => {
        dispatch(onChecking());
        try {
            const {data} = await calendarApi.post('/auth/new', { name, email, password });
            localStorage.setItem('token', data.token);
            localStorage.setItem('token-init-date', new Date().getTime().toString());
            dispatch(onLogin({name: data.name, _id: data.uid}))
        } catch(error) {
            const axiosError = error as AxiosError<{msg: string}>;
            dispatch(onLogout(axiosError.response?.data.msg || 'Error'))
            setTimeout(()=> {
                dispatch(clearErrorMesssage())
            }, 10)
        }
    }
    
    const checkAuthToken = async () => {
        const token = localStorage.getItem('token');
        if(!token) return dispatch(onLogout());
        
        try {
            const {data} = await calendarApi.get('/auth/renew');
            localStorage.setItem('token', data.token);
            localStorage.setItem('token-init-date', new Date().getTime().toString());
            dispatch(onLogin({name: data.name, _id: data.uid}))
        } catch(error) {
            localStorage.clear();
            dispatch(onLogout())
        }
    }

    const startLogout = () => {
        localStorage.clear();
        dispatch(onLogoutCalendar());
        dispatch(onLogout())
    }

    return{
        status,
        user,
        errorMessage,
        startLogin,
        startRegister,
        checkAuthToken,
        startLogout
    }
}