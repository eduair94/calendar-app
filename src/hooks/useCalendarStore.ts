import { useDispatch, useSelector } from "react-redux";
import { AuthUserI, RootState } from "../store";
import { CalendarEventI, onAddNewEvent, onDeleteEvent, onSetActiveEvent, onLoadEvents, onUpdateEvent, CalendarEventApiI } from "../store/calendar";
import { calendarApi } from "../api";
import { convertEventsToDateEvents } from "../helpers/converEventsToDateEvents";
import Swal from "sweetalert2";
import { AxiosError } from "axios";

export const useCalendarStore = () => {
    const dispatch = useDispatch();
    const {events, activeEvent} = useSelector((state:RootState) =>  state.calendar)
    const { user } = useSelector((state:RootState) =>  state.auth) as {user: AuthUserI}


    const setActiveEvent = (calendarEvent: CalendarEventI) => {
        dispatch(onSetActiveEvent(calendarEvent))
    }

    const startSavingEvent = async (calendarEvent: CalendarEventI) => {
        try {

            if(calendarEvent.id) {
                await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);
                dispatch(onUpdateEvent({...calendarEvent, user}))
                return;
            }

            const {data} = await calendarApi.post('/events', calendarEvent);
            const event:CalendarEventApiI = data.event;
            
            dispatch(onAddNewEvent({...calendarEvent, user, id: event.id}))

        } catch(e) {
            const axiosError = e as AxiosError<{msg:string}>;
            Swal.fire('Error', axiosError.response?.data.msg || 'Error saving event', 'error')
        }
    
    }

    const startDeletingEvent = async () => {
        try {
            await calendarApi.delete(`/events/${activeEvent?.id}`);
            dispatch(onDeleteEvent())
        } catch(e) {
            console.log('Error deleting event')
            console.log(e); 
        }
    }

    const startLoadingEvents = async () => {
        try {
            const {data} = await calendarApi.get('/events');
            const events = data.events || [];
            const eventsWithDate = convertEventsToDateEvents(events);
            dispatch(onLoadEvents(eventsWithDate))
        } catch(e) {
            console.log('Error loading events')
            console.log(e);

        }
    }

    return {
        setActiveEvent,
        events,
        activeEvent,
        startSavingEvent,
        startDeletingEvent,
        hasEventSelected: !!activeEvent,
        startLoadingEvents
    }
}