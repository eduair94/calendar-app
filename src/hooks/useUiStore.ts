import { useDispatch, useSelector } from "react-redux"
import { RootState, onCloseDateModal, onOpenDateModal } from "../store"

export const useUiStore = () => {
    const dispatch = useDispatch();
    const {isDateModalOpen} = useSelector((state:RootState) =>  state.ui)
    
    const openDateModal = () => {
        dispatch(onOpenDateModal());
    }

    const closeDateModal = () => {
        dispatch(onCloseDateModal())
    }
    const toggleDateModal = () => {
        if(isDateModalOpen)
            return closeDateModal();
        return openDateModal();
    }
    
    return {
        isDateModalOpen,
        openDateModal,
        closeDateModal,
        toggleDateModal
    }
}