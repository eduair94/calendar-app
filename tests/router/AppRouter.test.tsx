import { render, screen } from "@testing-library/react";
import React from "react";
import { AppRouter } from "../../src/router/AppRouter";
import { useAuthStore } from "../../src/hooks/useAuthStore";
import { AuthStatus } from "../../src/store";
import { MemoryRouter } from "react-router-dom";
jest.mock("../../src/hooks/useAuthStore");
jest.mock("react-modal", () => {
  return {
    setAppElement: jest.fn(),
    // other exports...
  };
});
jest.mock("../../src/calendar", () => ({
  CalendarPage: () => <div>CalendarPage</div>,
}));

describe("Tests in <AppRouter/>", () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it("should show the loading screen and call checkAuthToken ", () => {
    (useAuthStore as jest.Mock<unknown>).mockReturnValue({
      status: AuthStatus.checking,
      checkAuthToken: mockCheckAuthToken,
    });

    render(<AppRouter />);
    expect(screen.getByText("Cargando...")).toBeTruthy();
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  it("should show the login in case of not being authenticated", () => {
    (useAuthStore as jest.Mock<unknown>).mockReturnValue({
      status: AuthStatus.notAuthenticated,
      checkAuthToken: mockCheckAuthToken,
    });

    const { container } = render(
      <MemoryRouter initialEntries={["/auth/dfasfsd"]}>
        <AppRouter />
      </MemoryRouter>,
    );

    expect(screen.getAllByText("Login").length).toBeGreaterThanOrEqual(1);
    expect(container).toMatchSnapshot();
  });

  it("should show the calendar in case of being authenticated", () => {
    (useAuthStore as jest.Mock<unknown>).mockReturnValue({
      status: AuthStatus.authenticated,
      checkAuthToken: mockCheckAuthToken,
    });

    render(
      <MemoryRouter initialEntries={["/auth/dfasfsd"]}>
        <AppRouter />
      </MemoryRouter>,
    );

    expect(screen.getByText("CalendarPage")).toBeTruthy();
  });
});
