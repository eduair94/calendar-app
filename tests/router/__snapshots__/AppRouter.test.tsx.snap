// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <AppRouter/> should show the login in case of not being authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Login
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Register
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPasswordRepeat"
              placeholder="Repeat your password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
